lib/internal/readline/eventsToAsyncIteratorFactory.js

```diff
@@ -0,0 +1,133 @@
+'use strict';
+const {
+  ArrayPrototypeConcat,
+  Error,
+  Promise,
+  SymbolAsyncIterator,
+} = primordials;
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  },
+} = require('internal/errors');
+const FixedQueue = require('internal/fixed_queue');
+
+const kPauseThreshold = 1024;
+const kResumeThreshold = 1;
+const kItemEvents = ['data'];
+const kCloseEvents = ['close', 'end'];
+const kErrorEvents = ['error'];
+
+module.exports = function eventsToAsyncIteratorFactory(readable, {
```

Comment:
This looks pretty close to `on` from events and I am wondering if this can't be an extension to `on` (like "close on events") with on being made to support highWaterMark.

Then the implementation can be

```js
const iter = on(interface, 'line', { closeOn: ['close', 'end'], highWaterMark: 1024 });
return iter;
```

Or something similar?

Reply:
@benjamingr all right, that seems good! I'm working on it.
There're some differences between the two functions, though. For example, the on method yields array values, as the listened event can receive more than one parameter, while the async iterator created by eventsToAsyncIteratorFactory returns just a string, the expected line, but this can be solved by manipulating the resulting async iterator.

Another thing is that the on function uses an array to control the queue, which may or may not be slower than using the FixedQueue. I'm really not sure of it yet and maybe it'll be worth it to use the FixedQueue instead, but all of that I need to test very well to make sure that everything is ok.

I still didn't get it quite right but I'm working on it and I'll let it 100%, but I'd prefer to send that unification on another PR, if you don't mind, to not add that complexity in this one. If you think it's better to send it on this PR, though, no problem either, I'll let you know when I finish it.

Reply:
>  For example, the on method yields array values, as the listened event can receive more than one parameter, while the async iterator created by eventsToAsyncIteratorFactory returns just a string, the expected line, but this can be solved by manipulating the resulting async iterator.

I'm fine with either having `on` receive an optional private non-documented symbol to not emit an array or `map`ing it.

> Another thing is that the on function uses an array to control the queue, which may or may not be slower than using the FixedQueue.

It is likely slower and its performance would likely benefit from using the `FixedQueue` too, it has tests I think :] 

> If you think it's better to send it on this PR, though, no problem either

If it could be in this PR in two commits (one adding the performance optimizations here to `on` and another using `on` to do the change here + add the test) that would be optimal but I would also be fine with another PR for the unification and then updating this one to use that.

Reply:
@benjamingr all right. I finished the change you suggested and unified the functions, and I think it really worthed the effort.

First, I did it without changing the arrays used by the "on" function to FixedQueue, and just one test broke: the one I'm checking if the new async iterator for the readline got faster:

```
AssertionError [ERR_ASSERTION]: Failed: 60.18425208320601 isn't lesser than 81% of 65.32698930648154. Actual percentage: 92.13%
    at testPerformance (/home/thiago/projects/personal/node/test/parallel/test-readline-async-iterators.js:182:10)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
```
As you see, it got faster but not more than 19% faster, as I'm asserting.
So I changed it to use FixedQueue and all tests passed :).

So this evidence alone shows that changing the "on" function and using it on this PR can bring more benefits in general than before!

Thank you for pointing it out! It was really fun to do it!